### PR TITLE
Updated cpplint and cppcheck to exclude directories and files

### PR DESCRIPTION
**Pull request by ahcorde**
As we discussed in this PR https://github.com/ros2/geometry2/pull/258#issuecomment-621499030, it makes sense to exclude from cppcheck or cpplint some of the files or directories in a package.

We can use these arguments:
 - ament_cppcheck
```
  ament_cppcheck(LANGUAGE "c++"
    EXCLUDE_DIRS
      "include/tf2/LinearMath/* include/tf2/impl/*"
    EXCLUDE_FILES
       include/tf2/LinearMath/Scalar.h
       include/tf2/LinearMath/MinMax.h
  )
```
 - ament_lint
```
  ament_cpplint(EXCLUDE_DIRS include/tf2/LinearMath/)
```

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

**Comment by dirk-thomas** [comment]
There is no need for two separate arguments for files and directories. Also the excluded items should be configurable in CMake when using the automatic linting (by setting a global variable) where the package doesn't involve the CMake directly and can't pass a parameter.

See other linters for reference.

**Comment by ahcorde** [comment]
I added [language configuration](https://github.com/ahcorde/ament_lint/compare/ahcorde/cpp_linter_exclude...ahcorde:ahcorde/global_language?expand=1) to the global variables too for uncrustify and cppcheck. But I will open a PR when this is merged

**Comment by ahcorde** [comment]
Right now can be used:

```cmake
  set(ament_cmake_cppcheck_ADDITIONAL_EXCLUDE "include/tf2/LinearMath/*")
  set(ament_cmake_cpplint_ADDITIONAL_EXCLUDE_DIRS include/tf2/LinearMath/)
  set(ament_cmake_uncrustify_ADDITIONAL_ARGS --exclude LinearMath)

  find_package(ament_lint_auto REQUIRED)
  ament_lint_auto_find_test_dependencies()
```

The other PR will alllow:

```cmake
  set(ament_cmake_cppcheck_ADDITIONAL_LANGUAGE c++)
  set(ament_cmake_uncrustify_ADDITIONAL_LANGUAGE "C++")
```

File: ament_cmake_cppcheck/cmake/ament_cppcheck.cmake [hunk 1]
```diff
@@ -31,7 +33,7 @@
 # @public
 #
 function(ament_cppcheck)
-  cmake_parse_arguments(ARG "" "LANGUAGE;TESTNAME" "INCLUDE_DIRS" ${ARGN})
+  cmake_parse_arguments(ARG "" "LANGUAGE;TESTNAME;EXCLUDE" "INCLUDE_DIRS" ${ARGN})
```
Comment (dirk-thomas) [inline ament_cmake_cppcheck/cmake/ament_cppcheck.cmake:36]: Nitpick: aphabetical order.

Same below.

File: ament_cmake_cppcheck/cmake/ament_cppcheck.cmake [hunk 2]
```diff
@@ -47,6 +49,9 @@ function(ament_cppcheck)
   if(ARG_INCLUDE_DIRS)
     list(APPEND cmd "--include_dirs" "${ARG_INCLUDE_DIRS}")
   endif()
+  if(ARG_EXCLUDE)
+    list(APPEND cmd "--exclude" "${ARG_EXCLUDE}")
+  endif()
```
Comment (dirk-thomas) [inline ament_cmake_cppcheck/cmake/ament_cppcheck.cmake:54]: Nitpick: aphabetical order.

Same below.

File: ament_cpplint/ament_cpplint/main.py [hunk 3]
```diff
@@ -144,7 +148,14 @@ def main(argv=sys.argv[1:]):
...
         print('')
-        arguments += files
+        if args.excludedirs is not None:
+            filter_names = []
+            for f in files:
+                if any(exclude_dir not in f for exclude_dir in args.excludedirs):
+                    filter_names.append(f)
+            arguments += filter_names
+        else:
+            arguments += files
```
Comment (dirk-thomas) [inline ament_cpplint/ament_cpplint/main.py:158]: Please simplify the logic to a single `arguments += ...` with a list comprehension.

**Comment by ahcorde** [comment]
cppcheck allows directories and files

 - cmake: https://github.com/ahcorde/ament_lint/blob/ahcorde/cpp_linter_exclude/ament_cmake_cppcheck/cmake/ament_cppcheck.cmake#L26-L27
 - script. https://github.com/ahcorde/ament_lint/blob/ahcorde/cpp_linter_exclude/ament_cppcheck/ament_cppcheck/main.py#L69
 - hook: https://github.com/ahcorde/ament_lint/blob/ahcorde/cpp_linter_exclude/ament_cmake_cppcheck/cmake/ament_cmake_cppcheck_lint_hook.cmake#L38

cpplint allows directories

 - cmake: https://github.com/ahcorde/ament_lint/blob/ahcorde/cpp_linter_exclude/ament_cmake_cpplint/cmake/ament_cpplint.cmake#L29-L30
 - script: https://github.com/ahcorde/ament_lint/blob/ahcorde/cpp_linter_exclude/ament_cpplint/ament_cpplint/main.py#L81
 - hook: https://github.com/ahcorde/ament_lint/blob/ahcorde/cpp_linter_exclude/ament_cmake_cpplint/cmake/ament_cmake_cpplint_lint_hook.cmake#L30

**Comment by dirk-thomas** [comment]
> cpplint allows directories

In other API `cpplint` supports files too, e.g. https://github.com/ahcorde/ament_lint/blob/b156f37123d578765f80fd39fd328ba6f0539870/ament_cmake_cpplint/cmake/ament_cpplint.cmake#L31-L32

**Comment by ahcorde** [comment]
@dirk-thomas I allowed to exclude files too.

**Comment by dirk-thomas** [comment]
The file `cpplint.py` is imported from another project. Please do not modify it in this PR (since all changes are comments only).

**Comment by ahcorde** [comment]
Restored `cpplint.py`. Applied the patch in this PR https://github.com/ament/ament_lint/pull/238

**Comment by ahcorde** [comment]
Ruuning CI up-to `rcl`to check that linters (cppcheck, cpplint and uncrustify) are still working fine:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10709/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6121/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8722/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10608/)

File: ament_cppcheck/ament_cppcheck/main.py [hunk 4]
```diff
@@ -140,6 +145,8 @@ def main(argv=sys.argv[1:]):
         cmd.extend(['--language={0}'.format(args.language)])
     for include_dir in (args.include_dirs or []):
         cmd.extend(['-I', include_dir])
+    for exclude in (args.exclude or []):
+        cmd.extend(['--suppress=*:', exclude])
```
Comment (mm318) [inline ament_cppcheck/ament_cppcheck/main.py:149]: This code is wrong. This is effectively `cppcheck ... --suppress=*: <exclude> ...`, which appends `<exclude>` to the list of files to check but then ignores all checks on *all* files.

What you were looking for was actually `cppcheck ... --suppress=*:<exclude> ...` (without the space before `<exclude>`).

**Pull request merged**